examples/capture_circular.py

```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/python3
+
+from picamera2.encoders.h264_encoder import *
+from picamera2.picamera2 import *
+import time
+import os
+
+lsize = (320, 240)
+picam2 = Picamera2()
+video_config = picam2.video_configuration(main={"size": (1280, 720), "format": "RGB888"},
+                                          lores={"size": lsize, "format": "YUV420"})
+picam2.configure(video_config)
+picam2.start_preview()
+encoder = H264Encoder(1000000, repeat=True)
+encoder.output = CircularFileOutput()
+picam2.encoder = encoder
+picam2.start()
+picam2.start_encoder()
+
+w, h = lsize
+prev = None
+encoding = False
+ltime = 0
+
+while True:
+    cur = picam2.capture_buffer("lores")
+    cur = cur[:w * h].reshape(h, w)
+    if prev is not None:
+        # Measure pixels differences between current and
+        # previous frame
+        mse = np.square(np.subtract(cur, prev)).mean()
+        if mse > 7:
+            if not encoding:
+                epoch = int(time.time())
+                encoder.output.dumpbuffer("{}-before.h264".format(epoch), "{}.h264".format(epoch))
+                encoder.output.start()
```

Comment:
Yes, I think this is much improved now that the work is being done by the output object. We should perhaps consider also adding an example that uses the JPEG encoder as this would prove that we've done the output vs. encoder separation correctly.

On a purely aesthetic note, the "dumpbuffer" function sounds to me like a bit too much of the implementation is showing through in the application code. Could we perhaps overload the start function and call only that (with any necessary extra arguments)?

I can see there is some appeal in dumping the circular buffer to file here in the application thread, as it doesn't then block the camera and encode threads. I have some slight misgivings about thread safety here, but I won't digress now, perhaps I'll put them in a comment at the bottom of the general discussion.